File: clients/python/amenolink/_topic.py
from dataclasses import is_dataclass, asdict
from typing import Any, Callable, Generic
from ulid import ULID
from ._shared import AmenoException, T, _parse_data, client_config
from ._http_requests import _post_json
from ._connection_manager import connection_manager
from .dtos import Message, TopicMessage


class Topic(Generic[T]):
    def __init__(self, name: str, value_type: type[T] = Any):
        self.name: str = name
        self.value_type: type[T] = value_type
        self.disposed: bool = False
        self._handlers: set[Callable[[T | None, TopicMessage], None]] = set()

    def subscribe(self, handler: Callable[[T | None, TopicMessage], None]) -> None:
        self._ensure_not_disposed()
        self._handlers.add(handler)
        connection_manager.topic_manager.subscribe_topic(self)

    def publish(self, value: T | None = None, previous: Message | None = None) -> None:
        self._ensure_not_disposed()

        serialized_payload = value
        if hasattr(value, 'to_dict'):
            serialized_payload = value.to_dict()
        elif is_dataclass(value):
            serialized_payload = asdict(value)

        topic_message = TopicMessage(
            id=str(ULID()),
            previous=previous,
            topic=self.name,
            payload=serialized_payload,
            app_name=client_config.app_name,
        )

        url = f'{client_config.origin_url}/api/topic/publish'
        _post_json(url, topic_message.to_dict())

    def dispose(self) -> None:
        self._ensure_not_disposed()
        self.disposed = True
        self._handlers.clear()
        connection_manager.topic_manager.unsubscribe_topic(self)

    def _dispatch_message(self, message: TopicMessage) -> None:
        if self.disposed:
            return

        parsed_value = message.payload
        if self.value_type is not Any and message.payload is not None:
            parsed_value = _parse_data(message.payload, self.value_type)

        for handler_function in list(self._handlers):
            handler_function(parsed_value, message)

    def _ensure_not_disposed(self) -> None:
        if self.disposed:
            raise AmenoException(f"O tópico '{self.name}' já foi descartado (disposed).")


def topic(name: str, value_type: type[T] = Any) -> Topic[T]:
    return Topic(name=name, value_type=value_type)

File: clients/dart/amenolink/lib/src/topic.dart
import 'dart:async';
import 'package:ulid/ulid.dart';
import 'connection_manager.dart';
import 'dtos.dart';
import 'http_requests.dart';
import 'shared.dart';
import 'topic_manager.dart';

typedef TopicHandler<T> = void Function(TopicMessage<T> message);

class Topic<T> implements ITopic {
  @override
  final String name;
  bool _disposed = false;
  final Set<TopicHandler<T>> _handlers = {};

  Topic(this.name);

  void subscribe(TopicHandler<T> handler) {
    _ensureNotDisposed();
    _handlers.add(handler);
    connectionManager.topicManager.subscribeTopic(this);
  }

  Future<void> publish(T? value, {Message? previous}) async {
    _ensureNotDisposed();

    dynamic serializedPayload = value;
    if (value != null) {
      try {
        serializedPayload = (value as dynamic).toJson();
      } catch (_) {
        serializedPayload = value;
      }
    }

    final topicMessage = TopicMessage<dynamic>(
      id: Ulid().toString(),
      previous: previous,
      createdAt: DateTime.now().toUtc(),
      topic: name,
      payload: serializedPayload,
      appName: clientSetup.appName,
    );

    final url = '${clientSetup.originUrl}/api/topic/publish';
    await postJson(url, topicMessage.toJson());
  }

  void dispose() {
    _ensureNotDisposed();
    _disposed = true;
    _handlers.clear();
    connectionManager.topicManager.unsubscribeTopic(this);
  }

  @override
  void dispatchMessage(TopicMessage message) {
    if (_disposed) return;

    final rawPayload = message.payload;
    final parsedPayload = parseData<T>(rawPayload) as T?;

    final messageToDispatch = TopicMessage<T>(
      id: message.id,
      createdAt: message.createdAt,
      previous: message.previous,
      topic: message.topic,
      payload: parsedPayload,
      appName: message.appName,
    );

    for (final handlerFunction in List<TopicHandler<T>>.from(_handlers)) {
      handlerFunction(messageToDispatch);
    }
  }

  void _ensureNotDisposed() {
    if (_disposed) {
      throw AmenoException("O tópico '$name' já foi descartado (disposed).");
    }
  }
}

Topic<T> topic<T>(String name) => Topic<T>(name);
